src/context/authStore/authStore.tsx

```diff
@@ -0,0 +1,224 @@
+  clearAuthCredentials: () => void;
+}
+
+export const AuthStoreContext = createContext<AuthStoreContext>({
+  hasLoadedFromStore: false,
+  authCredentials: {},
+  setAuthCredentials: () => undefined,
+  removeAuthCredentials: () => undefined,
+  clearAuthCredentials: () => undefined,
+});
+
+/**
+ * Uses three versions of auth credential storage. Will automatically migrate to v3 and clear the values in v2 storage
+ * Assumes that no one uses v1 storage anymore as of Jul 2021.
+ *
+ * Versions:
+ *
+ * 1. (deprecated)
+ *    - Stores the corresponding values in AsyncStorage under the following keys: {@link SESSION_TOKEN_KEY}, {@link EXPIRY_KEY} and {@link ENDPOINT_KEY}
+ *    - Can only store one set of credentials
+ * 2. (deprecated)
+ *    - Stores a stringified {@link AuthCredentialsMap} in AsyncStorage under the key {@link AUTH_CREDENTIALS_STORE_KEY}
+ *    - Map from {{@link AuthCredentials.operatorToken}}{{@link AuthCredentials.endpoint}} to corresponding credentials
+ * 3. (current)
+ *    - Stores a stringified {@link AuthCredentialsMap} in SecureStorage under the key {@link AUTH_CREDENTIALS_STORE_KEY}
+ *    - Map from a hash of {{@link AuthCredentials.operatorToken}}{{@link AuthCredentials.endpoint}} to corresponding credentials
+ *    - Due to size limit of SecureStore, uses a bucketing method explained in {@link saveToStoreInBuckets}
+ */
+export const AuthStoreContextProvider: FunctionComponent<{
+  shouldMigrate?: boolean; // Temporary toggle to disable migration when testing
+}> = ({ shouldMigrate = true, children }) => {
+  /**
+   * Flag to mark when all loading has been done and consumers can read from the data.
+   * This includes any migration actions needed.
+   */
+  const [hasLoadedFromStore, setHasLoadedFromStore] = useState(false);
+  /**
+   * This flag marks when {@link authCredentials} is matching with the current state of
+   * the primary store. Any changes after this flag is true should trigger updates to the
+   * primary store.
+   *
+   * This flag should be set to true before any migration happens so that migrations will
+   * update the primary store.
+   *
+   * Documented in [#398](https://github.com/rationally-app/mobile-application/pull/398)
+   */
+  const [hasLoadedFromPrimaryStore, setHasLoadedFromPrimaryStore] = useState(
+    false
+  );
+  const [authCredentials, setAuthCredentialsMap] = useState<AuthCredentialsMap>(
+    {}
+  );
+  const prevAuthCredentials = usePrevious(authCredentials);
+
+  /** Effect to update the store when there are changes to authCredentials */
+  useEffect(() => {
+    if (hasLoadedFromPrimaryStore) {
+      const authCredentialsString = JSON.stringify(authCredentials);
+      const prevAuthCredentialsString = JSON.stringify(prevAuthCredentials);
+      // do a top level check to see if there are any changes
+      if (authCredentialsString !== prevAuthCredentialsString) {
+        // if there are changes, do a check for each bucket
+        saveToStoreInBuckets(
```

Comment:
I noticed we don't currently handle this promise rejecting. What happens if this rejects, e.g. the edge case for when the user has no more space?

I'm thinking we can log a sentry event so that we can monitor when saving to store fails. What do you think? Is there a need for some special handling for such errors?

Reply:
would there be any recovery from such an error?

Reply:
currently we don't await because the function `saveToStoreInBuckets` does not await internally either. Will depend on the above discussion https://github.com/rationally-app/mobile-application/pull/398#discussion_r678119850

Reply:
Yea we may not know the right recovery approach, could be lack of space, could be lack of permissions (?), etc.
To aid with knowing if this happens at all, could you catch the promise rejection and log the error to sentry?

Reply:
since our saving process is non-atomic, if there is an error, it might lead to corrupted data. is deleting the storage a valid recovery approach?

Reply:
It is indeed a valid recovery approach, though not a preferred one.